src/relay/qnn/op/convolution.cc

```diff
@@ -0,0 +1,434 @@
+  // Find which dimensions are C, R, S.
+  Array<Integer> axes_t3;
+  // For OIHW kernel layout, IHW are reduce axis
+  axes_t3 = {1, 2, 3};
+  if (param->kernel_layout == "HWIO") {
+    axes_t3 = {0, 1, 2};
+  }
+  auto reduced_t3 = Sum(Cast(weight, Int(32)), axes_t3, false, false);
+
+  // Find the newshape depenging on NCHW/NHWC layout.
+  Array<Integer> newshape;
+  // NCHW layout
+  newshape = {batch_size, out_channels, 1, 1};
+  if (param->data_layout == "NHWC") {
+    newshape = {batch_size, 1, 1, out_channels};
+  }
+  auto reshaped_t3 = Reshape(reduced_t3, newshape);
+  auto term3 = Multiply(zp_data, reshaped_t3);
+  return term3;
+}
+
+/*
+ * \brief Calculates the fourth term in the qnn.conv2d lowering sequence.
+ * \param param The qnn conv2d attributes.
+ * \param batch_size The batch size.
+ * \param in_channels The number of input channels.
+ * \param kernel_h The height of kernel.
+ * \param kernel_w The width of kernel.
+ * \return The sequence of Relay operatos for term4.
+ * \note The term4 looks like this
+ *
+ *       Sigma(c,r,s) zp_a * zp_w
+ *
+ */
+Expr Conv2DFourthTerm(const QConv2DAttrs* param, int batch_size, int in_channels, int kernel_h,
+                      int kernel_w) {
+  int scalar_term4 =
+      param->input_zero_point * param->kernel_zero_point * in_channels * kernel_h * kernel_w;
+  auto term4 = MakeConstantScalar(Int(32), scalar_term4);
+  return term4;
+}
+
+/*
+ * \brief Combines different terms of qnn conv2d lowering.
+ * \param term1 The term1 of qnn conv2d lowering.
+ * \param term2 The term2 of qnn conv2d lowering.
+ * \param term3 The term3 of qnn conv2d lowering.
+ * \param term4 The term4 of qnn conv2d lowering.
+ * \param param The qnn conv2d attributes.
+ * \return The combined sequence of relay operations.
+ * \note The combined operation looks like this
+ *
+ *       Sigma(c,r,s) QW(k, c, r, s) * QA(n, c, h + r, w + s)  // Term1
+ *     - Sigma(c,r,s) zp_w * QA(n, c, h + r, w + s)            // Term2
+ *     - Sigma(c,r,s) zp_a * QW(k, c, r, s)                    // Term3
+ *     + Sigma(c,r,s) zp_a * zp_w                              // Term4
+ *
+ */
+Expr Conv2DCombineTerms(const Expr& term1, const Expr& term2, const Expr& term3, const Expr& term4,
+                        const QConv2DAttrs* param) {
+  if (param->input_zero_point == 0 && param->kernel_zero_point == 0) {
+    // term 2, 3 and 4 become zero.
+    return term1;
+  } else if (param->input_zero_point == 0 && param->kernel_zero_point != 0) {
+    // term 3 and term 4 become zero.
+    return Subtract(term1, term2);
+  } else if (param->input_zero_point != 0 && param->kernel_zero_point == 0) {
+    // term 2 and term 4 become zero.
+    return Subtract(term1, term3);
+  } else {
+    auto data_term = Subtract(term1, term2);
+    // Putting constant terms together, so that constant folding can fold it.
+    auto const_term = Subtract(term4, term3);
+    return Add(data_term, const_term);
+  }
+}
+
+/*
```

Comment:
@FrozenGene @jackwish @u99127 

While the Requantize and Legalize pass are going through final details, it will be useful to prefetch and look at QNN conv2d lowering. Please review and let me know your comments.